[Edit] #824 - Remove EditableDomainObject

akhenry opened the pull request:
First pass at this, would like to collect some feedback on the approach used here.

In brief

- `EditableDomainObject` has been removed, it is no more. There is no navigation event when edit is initiated. This had some side-effects in `MctRepresentation`. This was because there is a navigation listener that triggers `refresh` on `MctRepresentation`. In order to minimize the side-effects of these changes, a status change listener has simply been added to `MctRepresentation` that makes a call to `Refresh`. Profiling suggests that the performance impact of this is minimal. Longer term, anything that requires refreshing on change into edit mode should use its own listeners instead of relying on MctRefresh to reload everything.
- There is a TransactionService that handles starting, committing, and canceling transactions. It also tracks dirty objects. For now objects are marked dirty on persist, as before. In the future dirtiness should be a consequence of mutation, not persistence.
- EditorCapability is now attached to all objects that support it automatically, not at edit time. This has been done so that the EditorCapability itself can be used to initiate edit (which encapsulates setting the edit flag, and initiating a transaction). It has the side effect of making `hasCapability("editor")` a test for editability of an object. Would suggest in this case we rename it `EditCapability`.
- Previously `EditableDomainObject` would wrap all capabilities so that they also returned `EditableDomainObjects`. Obviously this is no longer the case. For now, the editing object gets an `editing` status, but children do not. Consequently I have added an `inEditingContext` function to EditorCapability that will go up the context chain looking at the `editing` status. This replicates the prior situation, and allows for a simple drop-in code replacement for now. An alternative would be to decorate the CompositionCapability to set `editing` status on all children. This removes the immediate distinction between the object being edited, and those that are in just in the editing context, but this could be determined by using the navigated object instead.
- `EditorCapability` supports saving and canceling of an edit session. Save will commit the transaction, cancel will end it and refresh all dirty objects.

VWoeltjen commented on platform/commonUI/browse/src/creation/CreateAction.js:
```diff
@@ -88,22 +85,26 @@ define(
-            editableObject.setOriginalObject(parentObject);
-            editableObject.getCapability('status').set('editing', true);
-            editableObject.useCapability('mutation', function(model){
+            newObject.useCapability('mutation', function(model){
```
We could simplify slightly by setting `newModel.location = parentObject.getId()` before the `instantiation`

akhenry replied:
v. good point :smiley: Done.

VWoeltjen commented on platform/commonUI/browse/src/creation/CreateAction.js:
```diff
@@ -88,22 +85,26 @@ define(
 
-            if (countEditableViews(editableObject) > 0 && editableObject.hasCapability('composition')) {
-                this.navigationService.setNavigation(editableObject);
+            if (countEditableViews(newObject) > 0 && newObject.hasCapability('composition')) {
```
In the UI, we have an Edit button that either switches to Edit mode or shows an Edit Properties dialog. IIRC these are two different `edit` actions with same/similar policies. (With similar logic to this if.)

Are there tweaks to those `edit` actions that would allow us to replace this `if-else` with a single `perform('edit')`? (e.g. if they both return promises for success or failure...)

That said, it's a bit out of scope for this issue (is there another refactoring issue where this would be appropriate?)

akhenry replied:
Candidate for #628. Would like to separate out issues in order to keep changes to a minimum.

akhenry replied:
Deferring to #628. Agreed, would be nice to remove this ugly if statement

VWoeltjen commented on platform/commonUI/edit/src/actions/CancelAction.js:
```diff
@@ -82,7 +66,7 @@ define(
             var domainObject = (context || {}).domainObject;
             return domainObject !== undefined &&
-                domainObject.hasCapability("editor");
+                domainObject.getCapability("status").get("editing");
```
Now that `editor` is generally applicable (which I think is a good change) I think it would be useful to use this instead of the `editing` status for this kind of check. `domainObject.hasCapability('editor') && domainObject.getCapability('editor').isEditingContextRoot()`, for instance. (Believe this would solve the quirkiness of checking status sometimes, capabilities other times, which we've noticed before.)

VWoeltjen replied:
...looking down a little more, I think the _implementation_ of `isEditingContextRoot()` could just be a check for the `editing` status, for simplicity. Other code would still benefit from only needing to look at one capability instead of two.

akhenry replied:
@VWoeltjen in your mind, is there a semantic difference between `isEditingContextRoot()` and the function that I created `inEditingContext()`. If no, I'll rename it.

VWoeltjen replied:
Yes; the distinction I'm seeing:

1. `isEditingContextRoot`: Identifies the specific domain object used to initiate editing.
2. `inEditingContext`: Identifies any domain object that is part of the editing context (either because it is the specific domain object being edited, or because it is a sub-object being edited, e.g. in a timeline)

My reading was that the `editing` status was being used to distinguish the first case, and `inEditingContext` was being used to identify the more general second case. If they're actually identical, I would stick with the existing `inEditingContext` because it's simpler/clearer, but I would also use that check in lieu of checking status.

VWoeltjen commented on platform/commonUI/edit/src/capabilities/TransactionDecorator.js:
```diff
@@ -0,0 +1,68 @@
+    function (TransactionalPersistenceCapability) {
+        'use strict';
+
+        function TransactionDecorator(
```
I'd call this `TransactionCapabilityDecorator` to add some clarity about which service this decorates

akhenry replied:
Good point, renamed.

VWoeltjen commented on platform/commonUI/edit/src/capabilities/TransactionalPersistenceCapability.js:
```diff
@@ -0,0 +1,75 @@
+        TransactionalPersistenceCapability.prototype.persist = function () {
+            var domainObject = this.domainObject,
+                dirtyModelCache = this.dirtyModelCache;
+            if (this.transactionService.isActive() && !this.transactionService.isCommitting()) {
```
This is the only place these two methods (`isActive` and `isCommitting`) are used; could move the the `!committing` part into `isActive` itself to simplify this check (less knowledge of transactionService required to use it correctly, and it exposes a smaller interface)

akhenry replied:
Good point. Done.

VWoeltjen commented on platform/commonUI/edit/src/capabilities/TransactionalPersistenceCapability.js:
```diff
@@ -0,0 +1,75 @@
+                // from which 'catch' can be chained
+                return this.$q.when(true);
+            } else {
+                return this.persistenceCapability.persist().then(function (result) {
```
We should check the behavior here when persistence fails; a promise rejection would cause the `markClean` to be skipped.

But I'm not sure we need to mark anything clean at all here - if there's no active transaction, we're not in Edit mode and the object does not need to be marked clean, correct?

akhenry replied:
.persist() gets called when the transaction is committed. So in that case, the objects will have been marked dirty by previous persist calls. They need to be marked clean.

It was actually by design that markClean should only happen on successful persistence. This means that the dirty cache will be left populated with failed objects, which allows for the possibility of retries, and retains some useful information about the state of the object in memory.

VWoeltjen commented on platform/commonUI/edit/src/policies/EditableMovePolicy.js:
```diff
@@ -39,8 +39,8 @@ define([], function () {
-        if (key === 'move' && domainObject.hasCapability('editor')) {
-            return !!selectedObject && selectedObject.hasCapability('editor');
+        if (key === 'move' && domainObject.hasCapability('editor') && domainObject.getCapability('editor').inEditContext()) {
+            return !!selectedObject && selectedObject.hasCapability('editor') && selectedObject.getCapability('editor').inEditContext();
```
Line length of these conditions gets a little long, would like to break this up across multiple lines. Line breaks in an `if` condition are also a little funny-looking so maybe stick this in a function or assign it to a variable above?

akhenry replied:
Fixed

VWoeltjen commented on platform/commonUI/edit/src/services/TransactionService.js:
```diff
@@ -0,0 +1,118 @@
+
+        TransactionService.prototype.startTransaction = function () {
+            if (this.transaction)
+                console.error("Transaction already in progress")
```
Should use `$log.error` for consistency

akhenry replied:
Fixed.

VWoeltjen commented on platform/commonUI/edit/src/services/TransactionService.js:
```diff
@@ -0,0 +1,118 @@
+            }
+
+            function objectToPromise(object) {
+                return object.getCapability('persistence').persist();
```
If I'm understanding the flow right, the decorated `persist` gets called here, and doesn't just stick the object back in the dirty cache because `isCommitting()` is true at that point. This strikes me as a bit of a delicate balance between the three components (transaction service, persistence capability decorator, and dirty model queue)

What if, instead of getting objects from the queue, transactionService just let you push functions into it:

`transactionService.addToTransaction(onCommit, onCancel)`

When `commit` is called, it invokes all the `onCommits`; when `cancel` is called, it invokes all the `onCancels`. In either case, when it's done it clears the `active` flag and empties its queues. e.g.:

```
TransactionService.prototype.commit = function () {
    while (this.onCommits.length > 0) { // ...using a while in case some onCommit adds to transaction
        try { // ...also don't want to fail mid-loop...
            this.onCommits.unshift()();
        } catch (e) {
            this.$log.error("Error committing transaction.");
        }
    }
    this.active = false;
    this.onCommits = [];
    this.onCancels = [];
};
```

This would reduce TransactionService's dependencies greatly, making it much more testable, and less sensitive to any external changes.

akhenry replied:
I like the simplicity of what you're proposing. This did occur to me as well at the time, but I thought it might be useful to actually be able to track the dirtiness of objects, which is also why the dirty cache is implemented separately. Although we are not using dirtiness for anything other than transactions, this ability might be desirable in future.

On balance, I think your proposal makes sense, and I will refactor it. Dirtiness may be useful in future, but we will cross that bridge when we come to it.

VWoeltjen commented on platform/commonUI/edit/src/capabilities/TransactionalPersistenceCapability.js:
```diff
@@ -0,0 +1,75 @@
+        ) {
+            this.transactionService = transactionService;
+            this.dirtyModelCache = dirtyModelCache;
+            this.persistenceCapability = Object.create(persistenceCapability);
```
Not clear why there's an `Object.create` here

akhenry replied:
Not sure either, gone.

VWoeltjen commented on platform/commonUI/edit/src/services/TransactionService.js:
```diff
@@ -0,0 +1,118 @@
+            }
+
+            function objectToPromise(object) {
+                return self.$q.when(object.getModel().persisted && object.getCapability('persistence').refresh());
```
Think this `persisted` check should be moved into the `refresh` call itself

VWoeltjen commented on platform/representation/src/MCTRepresentation.js:
```diff
@@ -56,7 +56,8 @@ define(
             var representationMap = {},
-                gestureMap = {};
+                gestureMap = {},
+                listeners = 0;
```
Unused var?

akhenry replied:
Good catch, debugging code. Removed.

VWoeltjen commented:
Took a once over, added various comments in-line. Very nice to see such a large net decrease in lines of code - that alone demonstrates this is a much better approach.

akhenry commented:
* Addressed code review comments
* Added tests
* Re-enabled or replace related tests that had been disabled for NEM

### Author Checklist

1. Changes address original issue? __Y__
2. Unit tests included and/or updated with changes? __Y__
3. Command line build passes? __Y__
4. Changes have been smoke-tested? __Y__

akhenry commented:
Reassigning to address #905 and resolve new merge conflicts

akhenry commented:
> If I'm understanding the flow right, the decorated `persist` gets called here, and doesn't just stick the object back in the dirty cache because `isCommitting()` is true at that point. This strikes me as a bit of a delicate balance between the three components (transaction service, persistence capability decorator, and dirty model queue)

>What if, instead of getting objects from the queue, transactionService just let you push functions into it:

>`transactionService.addToTransaction(onCommit, onCancel)`

>When `commit` is called, it invokes all the `onCommits`; when `cancel` is called, it invokes all the `onCancels`. In either case, when it's done it clears the `active` flag and empties its queues. e.g.:

>```
TransactionService.prototype.commit = function () {
    while (this.onCommits.length > 0) { // ...using a while in case some onCommit adds to transaction
        try { // ...also don't want to fail mid-loop...
            this.onCommits.unshift()();
        } catch (e) {
            this.$log.error("Error committing transaction.");
        }
    }
    this.active = false;
    this.onCommits = [];
    this.onCancels = [];
};
```

>This would reduce TransactionService's dependencies significantly, making it much more testable, and less sensitive to any external changes.

I implemented this and it works really well, and reduce LOC and dependencies. I've hit a snag though. I replaced the dirty model tracking with a simple check of the modification date (which seems more correct anyway). This doesn't work where the child has been modified though, leading to #905. The obvious thing to do is to check the dirtiness state of all children at the same time, but this is an asynchronous process, meaning that it can't be used from a policy. The only place in our code that cares about dirtiness is the [EditNavigationPolicy](https://github.com/nasa/openmct/blob/master/platform/commonUI/edit/src/policies/EditNavigationPolicy.js), which is [invoked from the BrowseController](https://github.com/nasa/openmct/blob/master/platform/commonUI/browse/src/BrowseController.js#L88).

Three possibile solutions here
1. Do the navigation check in `BrowseController` on the navigated object __and all of its children__
2. On mutation, when the `modified` timestamp is updated, also update parent's modification date, and so on up the context chain
3. On mutation of an object, also set a 'dirty' flag on its parent, and so on up the context chain.

VWoeltjen commented on platform/commonUI/edit/src/capabilities/TransactionalPersistenceCapability.js:
```diff
@@ -0,0 +1,82 @@
+        TransactionalPersistenceCapability.prototype.persist = function () {
+            if (this.transactionService.isActive()) {
+                this.transactionService.addToTransaction(
+                    this.persistenceCapability.persist.bind(this.persistenceCapability),
```
Curious - if someone makes multiple edits to the same object during an editing session, does this result in multiple HTTP requests (e.g. when using ElasticSearch) when they Save?

If so, we should throttle these, or just use a `this.addedToTransaction` flag to make sure we only call `addToTransaction` once.

akhenry replied:
Ah, good point. I expect it would do, yes. I'll add a guard to prevent this happening.

VWoeltjen commented:
> I implemented this and it works really well, and has reduced LOC and dependencies. I've hit a snag though. I replaced the dirty model tracking with a simple check of the modification date (which seems more correct anyway). This doesn't work where the child has been modified though, leading to #905. The obvious thing to do is to check the dirtiness state of all children at the same time, but this is an asynchronous process, meaning that it can't be used from a policy. The only place in our code that cares about dirtiness is the EditNavigationPolicy, which is invoked from the BrowseController.

> Three possibile solutions here
1. Do the navigation check in BrowseController on the navigated object and all of its children
2. On mutation, when the modified timestamp is updated, also update parent's modification date, and so on up the context chain
3. On mutation of an object, also set a 'dirty' flag on its parent, and so on up the context chain.

Paraphrasing: EditorCapability#dirty used to return `true` whenever there were any unsaved changes (either to an object or its children); checking `modified` only returns `true` when the specific object has unsaved changes. So, the meaning of this method has changed.

Are there any cases where this change in meaning is necessary, or is it a side effect of removing the cache?

If it is just a side effect, I _think_ that adding a TransactionService#size method (returning the number of elements in the active transaction) and then reimplementing EditorCapability#dirty as `transactionService.size() > 0` would give results consistent with the former behavior. (Simply using `transactionService.isActive()` would also be correct in the majority of cases, but there's an edge case where this would return `true` in edit mode even if there were no changes at all.)

From a "what the code means" perspective this feels a _little_ off to me, but I think it basically makes sense - "there are tasks in a pending transaction" is a reasonable analogue of "there are unsaved changes". If we ever generalize the TransactionService for use outside of edit mode, we would likely need to revisit this, but that is already true for other reasons (and trying to generalize it at this point would be premature)

akhenry commented:
> Paraphrasing: EditorCapability#dirty used to return `true` whenever there were any unsaved changes (either to an object or its children); checking `modified` only returns `true` when the specific object has unsaved changes. So, the meaning of this method has changed.

> Are there any cases where this change in meaning is necessary, or is it a side effect of removing the cache?

The latter.

>If it is just a side effect, I _think_ that adding a TransactionService#size method (returning the number of elements in the active transaction) and then reimplementing EditorCapability#dirty as `transactionService.size() > 0` would give results consistent with the former behavior. (Simply using `transactionService.isActive()` would also be correct in the majority of cases, but there's an edge case where this would return `true` in edit mode even if there were no changes at all.)

Checking the size is a nice simple solution. It doesn't quite feel right, to my mind 'dirtiness' is about model state, rather than whether or not (potentially unnecessary) persist calls have been made. But it solves the problem quite succinctly.

Thanks

akhenry commented:
### Changes
* Added [size function](https://github.com/nasa/openmct/blob/open824_rebase/platform/commonUI/edit/src/services/TransactionService.js#L143) to TransactionService
* Changed [dirty check](https://github.com/nasa/openmct/blob/open824_rebase/platform/commonUI/edit/src/capabilities/EditorCapability.js#L114) on EditorCapability to use new `size()` function
* Added [guard code](https://github.com/nasa/openmct/blob/open824_rebase/platform/commonUI/edit/src/capabilities/TransactionalPersistenceCapability.js#L78) to TransactionalPersistenceCapability prevent multiple `persist()` calls for the same object
* Fixed performance issue in drag-drop composition positioning by changing [priority of decorator function](https://github.com/nasa/openmct/blob/open824_rebase/platform/commonUI/edit/bundle.js#L275)
* Added additional tests
* Fixed linting errors

### Author Checklist

1. Changes address original issue? __Y__
2. Unit tests included and/or updated with changes? __Y__
3. Command line build passes? __Y__
4. Changes have been smoke-tested? __Y__

akhenry commented:
@VWoeltjen There is another PR to consider along with this one - #931. This PR removes the status listener that I added in this PR, as well as all edit concerns from MctRepresentation. I've made it a separate PR so it can be considered in isolation. The reason that I added the status change listener on MctRepresentation was to force a refresh of all representations of an object when it switches to edit mode. This replicates the current situation where a navigation change will force a refresh of all representations. The reason for doing this was to minimize the delta from the current functionality and try and avoid subtle regressions due to capabilities, controllers etc. not getting refreshed.

In light of #931 though it makes sense to remove the edit logic from MctRepresentation, and in doing so the status listener also becomes redundant. As such I'd like your opinion on the side-effects of this change. The scope is actually somewhat limited because the browse-object representation (ie. the view) actually gets refreshed anyway due to a template change. Anything outside the scope of the view though might experience some strangeness (eg. tree, inspector). There was one case in the inspector where I have added the status listener to a controller to force a refresh.

VWoeltjen commented:
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

VWoeltjen commented:
Regarding #931, I think those changes are good and don't see any specific side effects, but let's merge against master instead of here. (Since there's some risk of regression there, want Revert to be easy.)

akhenry commented:
Agreed. Filing the PR against this branch was simply to make the diff a little easier to review. I'll file a new one against master.

akhenry commented:
Addresses #824
